R/interop-car.R

```diff
@@ -33,52 +33,96 @@ h_type2_contrast <- function(
-  contains_intercept <- (!0 %in% asg) &&
-    h_first_contain_categorical(effect, fcts, names(xlev))
-  coef_rows <- length(cols) - as.integer(contains_intercept)
-  l_mx <- matrix(0, nrow = coef_rows, ncol = length(asg))
+
+  categorical_covars <- intersect(rownames(fcts), names(xlev))
+
+  # An effect contains the model's intercept if the model was not built with an
+  # intercept and if the effect is the first one in the model specification to
+  # contain a categorical variable.
+  effect_contains_intercept <-
+    !attr(tms, "intercept") &&
+    length(categorical_covars) &&
+    effect == h_first_term_containing_categ(fcts, categorical_covars)
+
+  coef_rows <- length(cols) - as.integer(effect_contains_intercept)
+  l_mx <-
+    matrix(
+      0,
+      nrow = coef_rows,
+      ncol = ncol(mx),
+      dimnames = list(NULL, colnames(mx))
+    )
   if (coef_rows == 0L) {
+    l_mx <- l_mx[, !component(object, "beta_aliased"), drop = FALSE]
     return(l_mx)
   }
-  if (contains_intercept) {
-    l_mx[, cols] <- cbind(-1, diag(rep(1, coef_rows)))
-  } else {
-    l_mx[, cols] <- diag(rep(1, coef_rows))
-  }
+
+  # The effect's submatrix of l_mx shall simply be an identity matrix unless the
+  # effect contains the model's intercept. In this case, the submatrix has an
+  # extra column that should contain straight -1s.
+  l_mx[, cols] <-
+    if (effect_contains_intercept) {
+      cbind(diag(coef_rows), -1)
+    } else {
+      diag(coef_rows)
+    }
+
   for (i in setdiff(seq_len(ncol(fcts)), idx)) {
+    # This is where replacing the factors attributes' 2s with 1s matters.
     additional_vars <- names(which(fcts[, i] > fcts[, idx]))
-    additional_numeric <- any(!additional_vars %in% names(xlev))
+    additional_numeric <- !all(additional_vars %in% categorical_covars)
     current_col <- which(asg == i)
     if (
       ods[i] >= ods[idx] && all(fcts[, i] >= fcts[, idx]) && !additional_numeric
     ) {
+      # This is where it matters to use the complete design matrix including
+      # aliased columns.
       x1 <- mx[, cols, drop = FALSE]
       x0 <- mx[, -c(cols, current_col), drop = FALSE]
       x2 <- mx[, current_col, drop = FALSE]
-      m <- diag(rep(1, nrow(x0))) - x0 %*% solve(t(x0) %*% x0) %*% t(x0)
-      ret <- solve(t(x1) %*% m %*% x1) %*% t(x1) %*% m %*% x2
-      sub_mat <- if (contains_intercept) {
+      m <- diag(nrow(x0)) - h_quad_form_mat(x0, MASS::ginv(crossprod(x0)))
+      crossprod_x1_m <- crossprod(x1, m)
+      ret <- solve(crossprod_x1_m %*% x1) %*% crossprod_x1_m %*% x2
+      sub_mat <- if (effect_contains_intercept) {
         ret[-1, ] - ret[1, ]
       } else {
         ret
       }
       l_mx[, current_col] <- sub_mat
     }
   }
+  # Finally, we drop aliased columns. But we needed them to properly calculate
+  # the others.
+  l_mx <- l_mx[, !component(object, "beta_aliased"), drop = FALSE]
   l_mx[abs(l_mx) < tol] <- 0
   l_mx
 }
 
+h_first_term_containing_categ <- function(factors, categorical) {
```

Comment:
please add a little doc chunk above this function (we do this for all functions including internal, to help with maintenance later)